### PR TITLE
AA/MIN-256 Supply, borrow and net APY

### PR DESCRIPTION
**Pull request by ales-by**
**Description:**
Added RPC get_user_total_supply_borrow_and_net_apy for obtaining Borrow APY, Supply APY and Net APY
Describe in brief, what was done.

**Checklist**

- [x] Code is covered with tests
- [ ] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

**Comment by lut1k** [comment]
Can you make this function private?
https://github.com/minterest-finance/minterest-chain-node/blob/master/pallets/controller/src/lib.rs#L881

File: pallets/controller/src/lib.rs [hunk 1]
```diff
@@ -621,27 +624,18 @@ impl<T: Config> Pallet<T> {
...
-					pool_data.total_borrowed,
-				)?;
-
 				let oracle_price = T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;
 
-				let collateral_in_usd = Rate::from_inner(user_balance_wrapped_tokens)
-					.checked_mul(&current_exchange_rate)
-					.and_then(|x| x.checked_mul(&oracle_price))
+				let collateral_in_usd = Rate::from_inner(user_supply_balance)
+					.checked_mul(&oracle_price)
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:638]:
```suggestion
				let user_collateral_in_usd = Rate::from_inner(user_supply_balance)
					.checked_mul(&oracle_price)
```

File: pallets/controller/src/lib.rs [hunk 2]
```diff
@@ -663,11 +657,43 @@ impl<T: Config> Pallet<T> {
...
+			pool_data.total_borrowed,
+		)?;
+
+		let supply_balance = Rate::from_inner(user_balance_wrapped_tokens)
+			.checked_mul(&current_exchange_rate)
+			.map(|x| x.into_inner())
+			.ok_or(Error::<T>::NumOverflow)?;
+		Ok(supply_balance)
+	}
+
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:696]: This function differs only in name from `fn get_user_underlying_balance_per_asset()`. I guess there is no need to add the exact same function :)

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -693,6 +719,141 @@ impl<T: Config> Pallet<T> {
...
+				calculate_apy(net_apy_indicator, total_supply)?
+			}
+			net_apy_indicator if net_apy_indicator < Interest::zero() => {
+				calculate_apy(net_apy_indicator, total_borrow)?
+			}
+			_ => Interest::zero(),
+		};
+
+		Ok((supply_apy, borrow_apy, net_apy))
+	}
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:856]:
```suggestion
	/// Calculate total user's supply APY, borrow APY and Net APY.
	///
	/// - `who`: the AccountId whose APY should be calculated.
	pub fn get_user_total_supply_borrow_and_net_apy(
		who: T::AccountId,
	) -> Result<(Interest, Interest, Interest), DispatchError> {
		/*
		Annual Percentage Yield (APY) calculation:

		user_supply_interest = user_supply_usd * supply_rate;
		user_borrow_interest = user_borrow_usd * borrow_rate;
		user_mnt_supply_interest = user_supply_usd * mnt_supply_rate;
		user_mnt_borrow_interest = user_borrow_usd * mnt_borrow_rate;

		user_total_net_interest = Σ(user_supply_interest) - Σ(user_borrow_interest) +
		+ Σ(user_mnt_supply_interest) + Σ(user_mnt_borrow_interest);

		user_total_supply_APY = (Σ(user_supply_interest) / user_total_supply_usd) * blocks_per_year * 100, %
		user_total_borrow_APY = (Σ(user_borrow_interest) / user_total_borrow_usd) * blocks_per_year * 100, %
		user_total_net_APY:
			if user_total_net_interest > 0:
				(user_total_net_interest / user_total_supply_usd) * blocks_per_year * 100, %
			elif user_total_net_interest < 0:
				(user_total_net_interest / user_total_borrow_usd) * blocks_per_year * 100, %
		*/
		let calculate_interest = |amount: Balance, rate: Balance| {
			Interest::from_inner(amount as i128)
				.checked_mul(&Interest::from_inner(rate as i128))
				.ok_or(Error::<T>::NumOverflow)
		};

		let (
			user_total_supply_interest,
			user_total_borrow_interest,
			user_total_mnt_supply_interest,
			user_total_mnt_borrow_interest,
		) = CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset)
			.into_iter()
			.filter(|pool_id| T::LiquidityPoolsManager::pool_exists(pool_id))
			.try_fold(
				(Interest::zero(), Interest::zero(), Interest::zero(), Interest::zero()),
				|(
					acc_user_supply_interest,
					acc_user_borrow_interest,
					acc_user_mnt_supply_interest,
					acc_user_mnt_borrow_interest,
				),
				 pool_id|
				 -> result::Result<(Interest, Interest, Interest, Interest), DispatchError> {
					let user_supply_underlying = Self::get_user_supply_per_asset(&who, pool_id)?;
					let user_borrow_underlying = Self::get_user_borrow_per_asset(&who, pool_id)?;
					let oracle_price =
						T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;

					let user_supply_in_usd = Rate::from_inner(user_supply_underlying)
						.checked_mul(&oracle_price)
						.map(|x| x.into_inner())
						.ok_or(Error::<T>::BalanceOverflow)?;
					let user_borrow_in_usd = Rate::from_inner(user_borrow_underlying)
						.checked_mul(&oracle_price)
						.map(|x| x.into_inner())
						.ok_or(Error::<T>::BalanceOverflow)?;

					let (borrow_rate, supply_rate) =
						Self::get_liquidity_pool_borrow_and_supply_rates(pool_id).ok_or(Error::<T>::NumOverflow)?;
					let (mnt_borrow_rate, mnt_supply_rate) = T::MntManager::get_mnt_borrow_and_supply_rates(pool_id)?;

					let user_supply_interest = calculate_interest(user_supply_in_usd, supply_rate.into_inner())?;
					let user_borrow_interest = calculate_interest(user_borrow_in_usd, borrow_rate.into_inner())?;
					let user_mnt_supply_interest =
						calculate_interest(user_supply_in_usd, mnt_supply_rate.into_inner())?;
					let user_mnt_borrow_interest =
						calculate_interest(user_borrow_in_usd, mnt_borrow_rate.into_inner())?;
					Ok((
						acc_user_supply_interest
							.checked_add(&user_supply_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_borrow_interest
							.checked_add(&user_borrow_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_mnt_supply_interest
							.checked_add(&user_mnt_supply_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_mnt_borrow_interest
							.checked_add(&user_mnt_borrow_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
					))
				},
			)?;

		let user_net_interest = user_total_supply_interest
			.checked_sub(&user_total_borrow_interest)
			.and_then(|v| v.checked_add(&user_total_mnt_supply_interest))
			.and_then(|v| v.checked_add(&user_total_mnt_borrow_interest))
			.ok_or(Error::<T>::BalanceOverflow)?;

		let (user_total_supply_usd, user_total_borrow_usd) = Self::get_total_supply_and_borrowed_usd_balance(&who)?;

		// Calculate APY given the amount of blocks_per_year.
		let calculate_apy = |interest: Interest, amount: Balance| {
			interest
				.checked_div(&Interest::from_inner(amount as i128))
				.and_then(|v| v.checked_mul(&Interest::saturating_from_integer(T::BlocksPerYear::get())))
				.ok_or(Error::<T>::NumOverflow)
		};

		let user_total_supply_apy = calculate_apy(user_total_supply_interest, user_total_supply_usd)?;
		let user_total_borrow_apy = calculate_apy(user_total_borrow_interest, user_total_borrow_usd)?;
		let user_net_apy = match user_net_interest.is_positive() {
			true => calculate_apy(user_net_interest, user_total_supply_usd)?,
			false => calculate_apy(user_net_interest, user_total_borrow_usd)?,
		};

		Ok((user_total_supply_apy, user_total_borrow_apy, user_net_apy))
	}
```

Comment (lut1k) [inline pallets/controller/src/lib.rs:856]:
```suggestion
	/// Calculate total user's supply APY, borrow APY and Net APY.
	///
	/// - `who`: the AccountId whose APY should be calculated.
	pub fn get_user_total_supply_borrow_and_net_apy(
		who: T::AccountId,
	) -> Result<(Interest, Interest, Interest), DispatchError> {
		/*
		Annual Percentage Yield (APY) calculation:

		user_supply_interest = user_supply_usd * supply_rate;
		user_borrow_interest = user_borrow_usd * borrow_rate;
		user_mnt_supply_interest = user_supply_usd * mnt_supply_rate;
		user_mnt_borrow_interest = user_borrow_usd * mnt_borrow_rate;

		user_total_net_interest = Σ(user_supply_interest) - Σ(user_borrow_interest) +
		+ Σ(user_mnt_supply_interest) + Σ(user_mnt_borrow_interest);

		user_total_supply_APY = (Σ(user_supply_interest) / user_total_supply_usd) * blocks_per_year * 100, %
		user_total_borrow_APY = (Σ(user_borrow_interest) / user_total_borrow_usd) * blocks_per_year * 100, %
		user_total_net_APY:
			if user_total_net_interest > 0:
				(user_total_net_interest / user_total_supply_usd) * blocks_per_year * 100, %
			elif user_total_net_interest < 0:
				(user_total_net_interest / user_total_borrow_usd) * blocks_per_year * 100, %
		*/
		let calculate_interest = |amount: Balance, rate: Balance| {
			Interest::from_inner(amount as i128)
				.checked_mul(&Interest::from_inner(rate as i128))
				.ok_or(Error::<T>::NumOverflow)
		};

		let (
			user_total_supply_interest,
			user_total_borrow_interest,
			user_total_mnt_supply_interest,
			user_total_mnt_borrow_interest,
		) = CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset)
			.into_iter()
			.filter(|pool_id| T::LiquidityPoolsManager::pool_exists(pool_id))
			.try_fold(
				(Interest::zero(), Interest::zero(), Interest::zero(), Interest::zero()),
				|(
					acc_user_supply_interest,
					acc_user_borrow_interest,
					acc_user_mnt_supply_interest,
					acc_user_mnt_borrow_interest,
				),
				 pool_id|
				 -> result::Result<(Interest, Interest, Interest, Interest), DispatchError> {
					let user_supply_underlying = Self::get_user_supply_per_asset(&who, pool_id)?;
					let user_borrow_underlying = Self::get_user_borrow_per_asset(&who, pool_id)?;
					let oracle_price =
						T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;

					let user_supply_in_usd = Rate::from_inner(user_supply_underlying)
						.checked_mul(&oracle_price)
						.map(|x| x.into_inner())
						.ok_or(Error::<T>::BalanceOverflow)?;
					let user_borrow_in_usd = Rate::from_inner(user_borrow_underlying)
						.checked_mul(&oracle_price)
						.map(|x| x.into_inner())
						.ok_or(Error::<T>::BalanceOverflow)?;

					let (borrow_rate, supply_rate) =
						Self::get_liquidity_pool_borrow_and_supply_rates(pool_id).ok_or(Error::<T>::NumOverflow)?;
					let (mnt_borrow_rate, mnt_supply_rate) = T::MntManager::get_mnt_borrow_and_supply_rates(pool_id)?;

					let user_supply_interest = calculate_interest(user_supply_in_usd, supply_rate.into_inner())?;
					let user_borrow_interest = calculate_interest(user_borrow_in_usd, borrow_rate.into_inner())?;
					let user_mnt_supply_interest =
						calculate_interest(user_supply_in_usd, mnt_supply_rate.into_inner())?;
					let user_mnt_borrow_interest =
						calculate_interest(user_borrow_in_usd, mnt_borrow_rate.into_inner())?;
					Ok((
						acc_user_supply_interest
							.checked_add(&user_supply_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_borrow_interest
							.checked_add(&user_borrow_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_mnt_supply_interest
							.checked_add(&user_mnt_supply_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_mnt_borrow_interest
							.checked_add(&user_mnt_borrow_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
					))
				},
			)?;

		let user_net_interest = user_total_supply_interest
			.checked_sub(&user_total_borrow_interest)
			.and_then(|v| v.checked_add(&user_total_mnt_supply_interest))
			.and_then(|v| v.checked_add(&user_total_mnt_borrow_interest))
			.ok_or(Error::<T>::BalanceOverflow)?;

		let (user_total_supply_usd, user_total_borrow_usd) = Self::get_total_supply_and_borrowed_usd_balance(&who)?;

		// Calculate APY given the amount of blocks_per_year.
		let calculate_apy = |interest: Interest, amount: Balance| {
			interest
				.checked_div(&Interest::from_inner(amount as i128))
				.and_then(|v| v.checked_mul(&Interest::saturating_from_integer(T::BlocksPerYear::get())))
				.ok_or(Error::<T>::NumOverflow)
		};

		let user_total_supply_apy = calculate_apy(user_total_supply_interest, user_total_supply_usd)?;
		let user_total_borrow_apy = calculate_apy(user_total_borrow_interest, user_total_borrow_usd)?;
		let user_net_apy = match user_net_interest.is_positive() {
			true => calculate_apy(user_net_interest, user_total_supply_usd)?,
			false => calculate_apy(user_net_interest, user_total_borrow_usd)?,
		};

		Ok((user_total_supply_apy, user_total_borrow_apy, user_net_apy))
	}
```

Comment (lut1k) [reply pallets/controller/src/lib.rs:856]: 1. I didn't like the idea of ​​a separate function `fn get_vec_of_supply_and_borrow_balance ()`. In your implementation, you use three cycles `for` for calculations, in my opinion this is redundant.
2. I proposed a new implementation based on the following considerations:
- I use ready-made functions: `fn get_user_supply_per_asset()`, `fn get_user_borrow_per_asset()`, `fn get_liquidity_pool_borrow_and_supply_rates()`, `fn get_mnt_borrow_and_supply_rates()`, `fn get_total_supply_and_borrowed_usd_balance()`. All these functions within themselves contain a call to the same function `fn calculate_current_pool_data ()` in order to calculate the current pool data, that is, with the current implementation, we call the function `fn calculate_current_pool_data ()` 20 times !!!. I propose now to call these functions directly, and in the future to refactor, namely: write separate internal functions, for example `fn get_user_supply_per_asset_inner (pool_state)`, these internal functions will receive the already calculated current state of the pool as an argument.
- or do not use ready-made functions and manually call `fn calculate_current_pool_data ()` and calculate all parameters. I don't really like this variant.

File: pallets/controller/src/lib.rs [hunk 4]
```diff
@@ -693,6 +719,141 @@ impl<T: Config> Pallet<T> {
...
+					vec.push((pool_id, supply_balance, borrow_balance));
+					Ok(vec)
+				},
+			)
+	}
+
+	/// Calculate total user's supply APY, borrow APY and Net APY.
+	///
+	/// - `who`: the AccountId whose APY should be calculated.
+	pub fn get_user_total_supply_borrow_and_net_apy(
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:766]:
```suggestion
	/// Calculate total user's supply APY, borrow APY and Net APY.
	///
	/// - `who`: the AccountId whose APY should be calculated.
	pub fn get_user_total_supply_borrow_and_net_apy(
		who: T::AccountId,
	) -> Result<(Interest, Interest, Interest), DispatchError> {
		/*
		Annual Percentage Yield (APY) calculation:

		user_supply_interest = user_supply_usd * supply_rate;
		user_borrow_interest = user_borrow_usd * borrow_rate;
		user_mnt_supply_interest = user_supply_usd * mnt_supply_rate;
		user_mnt_borrow_interest = user_borrow_usd * mnt_borrow_rate;

		user_total_net_interest = Σ(user_supply_interest) - Σ(user_borrow_interest) +
		+ Σ(user_mnt_supply_interest) + Σ(user_mnt_borrow_interest);

		user_total_supply_APY = (Σ(user_supply_interest) / user_total_supply_usd) * blocks_per_year * 100, %
		user_total_borrow_APY = (Σ(user_borrow_interest) / user_total_borrow_usd) * blocks_per_year * 100, %
		user_total_net_APY:
			if user_total_net_interest > 0:
				(user_total_net_interest / user_total_supply_usd) * blocks_per_year * 100, %
			elif user_total_net_interest < 0:
				(user_total_net_interest / user_total_borrow_usd) * blocks_per_year * 100, %
		*/
		let calculate_interest = |amount: Balance, rate: Balance| {
			Interest::from_inner(amount as i128)
				.checked_mul(&Interest::from_inner(rate as i128))
				.ok_or(Error::<T>::NumOverflow)
		};

		let (
			user_total_supply_interest,
			user_total_borrow_interest,
			user_total_mnt_supply_interest,
			user_total_mnt_borrow_interest,
		) = CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset)
			.into_iter()
			.filter(|pool_id| T::LiquidityPoolsManager::pool_exists(pool_id))
			.try_fold(
				(Interest::zero(), Interest::zero(), Interest::zero(), Interest::zero()),
				|(
					acc_user_supply_interest,
					acc_user_borrow_interest,
					acc_user_mnt_supply_interest,
					acc_user_mnt_borrow_interest,
				),
				 pool_id|
				 -> result::Result<(Interest, Interest, Interest, Interest), DispatchError> {
					let user_supply_underlying = Self::get_user_supply_per_asset(&who, pool_id)?;
					let user_borrow_underlying = Self::get_user_borrow_per_asset(&who, pool_id)?;
					let oracle_price =
						T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;

					let user_supply_in_usd = Rate::from_inner(user_supply_underlying)
						.checked_mul(&oracle_price)
						.map(|x| x.into_inner())
						.ok_or(Error::<T>::BalanceOverflow)?;
					let user_borrow_in_usd = Rate::from_inner(user_borrow_underlying)
						.checked_mul(&oracle_price)
						.map(|x| x.into_inner())
						.ok_or(Error::<T>::BalanceOverflow)?;

					let (borrow_rate, supply_rate) =
						Self::get_liquidity_pool_borrow_and_supply_rates(pool_id).ok_or(Error::<T>::NumOverflow)?;
					let (mnt_borrow_rate, mnt_supply_rate) = T::MntManager::get_mnt_borrow_and_supply_rates(pool_id)?;

					let user_supply_interest = calculate_interest(user_supply_in_usd, supply_rate.into_inner())?;
					let user_borrow_interest = calculate_interest(user_borrow_in_usd, borrow_rate.into_inner())?;
					let user_mnt_supply_interest =
						calculate_interest(user_supply_in_usd, mnt_supply_rate.into_inner())?;
					let user_mnt_borrow_interest =
						calculate_interest(user_borrow_in_usd, mnt_borrow_rate.into_inner())?;
					Ok((
						acc_user_supply_interest
							.checked_add(&user_supply_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_borrow_interest
							.checked_add(&user_borrow_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_mnt_supply_interest
							.checked_add(&user_mnt_supply_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
						acc_user_mnt_borrow_interest
							.checked_add(&user_mnt_borrow_interest)
							.ok_or(Error::<T>::BalanceOverflow)?,
					))
				},
			)?;

		let user_net_interest = user_total_supply_interest
			.checked_sub(&user_total_borrow_interest)
			.and_then(|v| v.checked_add(&user_total_mnt_supply_interest))
			.and_then(|v| v.checked_add(&user_total_mnt_borrow_interest))
			.ok_or(Error::<T>::BalanceOverflow)?;

		let (user_total_supply_usd, user_total_borrow_usd) = Self::get_total_supply_and_borrowed_usd_balance(&who)?;

		// Calculate APY given the amount of blocks_per_year.
		let calculate_apy = |interest: Interest, amount: Balance| {
			interest
				.checked_div(&Interest::from_inner(amount as i128))
				.and_then(|v| v.checked_mul(&Interest::saturating_from_integer(T::BlocksPerYear::get())))
				.ok_or(Error::<T>::NumOverflow)
		};

		let user_total_supply_apy = calculate_apy(user_total_supply_interest, user_total_supply_usd)?;
		let user_total_borrow_apy = calculate_apy(user_total_borrow_interest, user_total_borrow_usd)?;
		let user_net_apy = match user_net_interest.is_positive() {
			true => calculate_apy(user_net_interest, user_total_supply_usd)?,
			false => calculate_apy(user_net_interest, user_total_borrow_usd)?,
		};

		Ok((user_total_supply_apy, user_total_borrow_apy, user_net_apy))
	}
```
I didn't like the idea of   a separate function fn get_vec_of_supply_and_borrow_balance (). In your implementation, you use three cycles for for calculations, in my opinion this is redundant.
I proposed a new implementation based on the following considerations:
I use ready-made functions: fn get_user_supply_per_asset(), fn get_user_borrow_per_asset(), fn get_liquidity_pool_borrow_and_supply_rates(), fn get_mnt_borrow_and_supply_rates(), fn get_total_supply_and_borrowed_usd_balance(). All these functions within themselves contain a call to the same function fn calculate_current_pool_data () in order to calculate the current pool data, that is, with the current implementation, we call the function fn calculate_current_pool_data () 20 times !!!. I propose now to call these functions directly, and in the future to refactor, namely: write separate internal functions, for example fn get_user_supply_per_asset_inner (pool_state), these internal functions will receive the already calculated current state of the pool as an argument.
or do not use ready-made functions and manually call fn calculate_current_pool_data () and calculate all parameters. I don't really like this variant.

**Review by Jekshmek — DISMISSED** [review]
I like how you did

File: primitives/src/lib.rs [hunk 5]
```diff
@@ -71,6 +71,9 @@ pub use sp_runtime::OpaqueExtrinsic as UncheckedExtrinsic;
 /// An instant or duration in time.
 pub type Moment = u64;
 
+/// Decimal representation of interest. Signed.
+pub type Interest = FixedI128;
+
```
[the remark this answers is not in the document]
Comment (ales-by) [reply primitives/src/lib.rs:76]: fixed

File: pallets/controller/src/lib.rs (hunk 1 above)
Comment (ales-by) [reply pallets/controller/src/lib.rs]: fixed

**Review by Jekshmek — DISMISSED** [review]

File: pallets/controller/src/lib.rs (hunk 2 above)
Comment (ales-by) [reply pallets/controller/src/lib.rs:696]: fixed

File: pallets/controller/src/lib.rs [hunk 6]
```diff
@@ -693,6 +719,141 @@ impl<T: Config> Pallet<T> {
...
+						Interest::from_inner(amount as i128)
+							.checked_mul(&Interest::from_inner(rate as i128))
+							.ok_or(Error::<T>::NumOverflow)
+					};
+
+					let supply_interest = calculate_interest(current_supply_in_usd, supply_rate.into_inner())?;
+					let borrow_interest = calculate_interest(current_borrow_in_usd, borrow_rate.into_inner())?;
+
+					// Calculate mnt interest for both borrow and supply.
+					let (mnt_borrow_rate, mnt_supply_rate) = T::MntManager::get_mnt_borrow_and_supply_rates(pool_id)?;
```
[the remark this answers is not in the document]
Comment (ales-by) [reply pallets/controller/src/lib.rs:809]: Also thought about this.
@Apollo50 had some explanations why it is ok, if I don't miss something...

File: pallets/controller/src/lib.rs (hunk 4 above)
Comment (ales-by) [reply pallets/controller/src/lib.rs:766]: Don't like ` fn get_vec_of_supply_and_borrow_balance()` as well. 
I'll use your realization with some minor changes.

And it is a very good point about kind of cashing `fn calculate_current_pool_data()` result. We'll need to think about this later.

File: pallets/controller/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs]: I would like to hear these explanations)

[the remark this answers is not in the document]
Comment (Apollo50) [reply pallets/controller/src/lib.rs]: Before updating the logic of the mnt pallets, there were difficulties with updating speeds with a fresh value of the total borrow. Now there is no such problem and you can return fresh indices without any problems.

[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs]: We discussed this situation. We have changed the logic of working with MNT speeds. Therefore, now we have the opportunity to implement the calculation of the current values, and not the old ones.
call `Controller::calculate_current_pool_data()` via `LiquidityPoolsManager::get_pool_total_borrowed()` and etc..
To do this, you need to make the function `fn calculate_current_pool_data()` public)

**Review by Jekshmek — APPROVED** [review]

**Review by Jekshmek — APPROVED** [review]

File: pallets/controller/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ales-by) [reply pallets/controller/src/lib.rs]: MNT rate actualization is moved into MIN-472 [TECH DEBT] Fix MNT rates calculation to return fresh values

**Review by Jekshmek — APPROVED** [review]

File: pallets/traits/src/lib.rs [hunk 7]
```diff
@@ -212,6 +212,9 @@ pub trait MinterestModelManager {
 		underlying_asset: CurrencyId,
 		utilization_rate: Rate,
 	) -> Result<Rate, DispatchError>;
+
+	/// Returns BlocksPerYear value
+	fn get_blocks_per_year() -> u128;
```
Comment (lut1k) [inline pallets/traits/src/lib.rs:217]: In my opinion, it is better to just use import from the constants module:
`use minterest_primitives::constants::time::{BLOCKS_PER_YEAR, DAYS};`

**Review by Jekshmek — DISMISSED** [review]

**Review by lut1k — APPROVED** [review]

File: pallets/controller/src/lib.rs [hunk 8]
```diff
@@ -634,6 +641,131 @@ impl<T: Config> Pallet<T> {
...
+				},
+			)?;
+
+		let user_net_interest = user_total_supply_interest
+			.checked_sub(&user_total_borrow_interest)
+			.and_then(|v| v.checked_add(&user_total_mnt_supply_interest))
+			.and_then(|v| v.checked_add(&user_total_mnt_borrow_interest))
+			.ok_or(Error::<T>::BalanceOverflow)?;
+
+		let (user_total_supply_usd, user_total_borrow_usd) =
```
Comment (Apollo50) [inline pallets/controller/src/lib.rs:744]: Why did you decide not to calculate total_* values in loop above ?
There are everything you need to calculate it

File: runtime/src/rpc_tests.rs [hunk 9]
```diff
@@ -1606,3 +1610,133 @@ fn protocol_interest_transfer_should_work() {
...
+			);
+
+			// ETH mnt borrow:  5 * 4 / (80_000 * 2) = 0.000125
+			// ETH mnt supply:  5 * 4 / (100_000 * 2) = 0.0001
+			assert_eq!(
+				get_mnt_borrow_and_supply_rates(ETH),
+				(Rate::from_inner(125000000000000), Rate::from_inner(100000000000000))
+			);
+
+			// MNT interest for 1 block
```
Comment (Apollo50) [inline runtime/src/rpc_tests.rs:1691]: For 1 year mb?

File: pallets/traits/src/lib.rs (hunk 7 above)
Comment (ales-by) [reply pallets/traits/src/lib.rs]: Sure. Fixed.

**Review by lut1k — DISMISSED** [review]

**Review by Apollo50 — APPROVED** [review]

**Review by lut1k — APPROVED** [review]

**Review by Jekshmek — APPROVED** [review]

**Pull request merged**